### PR TITLE
Caching k8sclient

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/16406

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Comment by tomahawk28** [comment]
I'm testing this from `ghost`. I will check tomorrow morning to see whether we have a problem with a cache.

File: service/internal/k8sclient/cache.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	if v, ok := r.cache.Get(k); ok {
+		clients, ok := v.(k8sclient.Interface)
+		if !ok {
+			return nil, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", k8sclient.Clients{}, v)
+		}
+
+		return clients, nil
+	}
+
+	fmt.Println("CACHE FAULT")
```
Comment (tomahawk28) [inline service/internal/k8sclient/cache.go:87]: TODO: remove these right before merge

File: service/internal/k8sclient/cache.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	fmt.Println("CREATE CLIENT")
+	clients, err := r.generateK8sClient(ctx, kubeConfig)
```
Comment (tomahawk28) [inline service/internal/k8sclient/cache.go:89]: Only creating the k8sclient when we have a cache fault.

File: service/internal/k8sclient/cache.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	fmt.Println("CREATE KEY")
+	r.cache.SetDefault(k, clients)
```
Comment (tomahawk28) [inline service/internal/k8sclient/cache.go:95]: I'm not sure TTL of each k8sclient, I left the testing operator overnight in `ghost` but it's okay.  Does that means it have more than 12 hours as TTL?

File: service/controller/app/app.go [hunk 4]
```diff
@@ -17,14 +17,16 @@ import (
 	"github.com/giantswarm/app-operator/v4/pkg/label"
 	"github.com/giantswarm/app-operator/v4/pkg/project"
 	"github.com/giantswarm/app-operator/v4/service/controller/app/controllercontext"
+	cachedk8sclient "github.com/giantswarm/app-operator/v4/service/internal/k8sclient"
```
Comment (rossf7) [inline service/controller/app/app.go:20]:
```suggestion
	"github.com/giantswarm/app-operator/v4/service/internal/k8sclientcache"
```

I found the naming a bit hard to follow here. What about this?

File: service/internal/k8sclient/cache.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/giantswarm/kubeconfig/v4"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	gocache "github.com/patrickmn/go-cache"
+	"github.com/spf13/afero"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	expiration = 1 * time.Hour
```
Comment (rossf7) [inline service/internal/k8sclient/cache.go:20]: I think this TTL is a bit long. The problem is cert renewal for the app-operator cert as I think then the operator won't be able to connect.

What about reducing it to 10 mins or even 5 mins?

**Review by rossf7 — COMMENTED** [review]
Looks good! Just a couple of queries.

File: service/controller/app/app.go (hunk 4 above)
Comment (tomahawk28) [reply service/controller/app/app.go:20]: Much better!

File: service/internal/k8sclient/cache.go (hunk 5 above)
Comment (tomahawk28) [reply service/internal/k8sclient/cache.go:20]: I think reducing it to 10 mins makes more sense.

File: service/internal/k8sclientcache/cache.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	"k8s.io/client-go/rest"
+)
+
+const (
+	expiration = 10 * time.Minute
+)
+
+type Config struct {
+	// Dependencies.
+	Fs        afero.Fs
```
Comment (rossf7) [inline service/internal/k8sclientcache/cache.go:25]: What do we need the filesystem for?

**Review by rossf7 — APPROVED** [review]
LGTM 

Once debug is removed and let's also remove the filesystem unless its needed.

File: service/internal/k8sclient/cache.go (hunk 5 above)
Comment (rossf7) [reply service/internal/k8sclient/cache.go]: Cool thanks 👍

File: service/internal/k8sclientcache/cache.go (hunk 6 above)
Comment (tomahawk28) [reply service/internal/k8sclientcache/cache.go:25]: Good catch, I coped it from `client` resource but we don't need it as you said.
